### PR TITLE
Upgrade to Gradle 6.7 and use toolchains feature to build on 11 and t…

### PR DESCRIPTION
**Pull request by anuraaga**
…est on 11/8.

Gradle 6.7 has a great feature to automatically download toolchains when it can't find them on the system. This can allow easily testing multiple Java versions without build system tweaks. As long as the user has any version of Java installed, Gradle will run, and it will automatically downloads additional needed toolchains, including the primary one (e.g. Java 8 user can run our build and it will use Java 11)

This tweaks the build so now

- Build always uses Java 11. In the future, I would probably bump this to 15 - while runtime tweaks are often backported to older Java, compiler tweaks rarely are. I'd generally recommend using latest Java for building to get the latest compiler tweaks (which are in practice mostly improvements to Javadoc), when targeted to older Java versions and running tests on them there's basically no downside
  - `release` flag is used instead of source/target compatibility to allow Java to enforce classpath version

- Tests default to using Java 11. With the `-PtestAdditionalJavaVersions`, additionally `testJava8` task runs the tests using Java 8. CI enables this flag on ubuntu test runner in GitHub and CircleCI

### CODE REVIEW

**Comment by anuraaga** [comment]
While debugging CircleCI I noticed we've never built on Java 11 using it due to `executorType: machine` XD Murphy's law of course prevails as now that I thought we can finally delete it, a test flaked on GitHub

**Comment by anuraaga** [comment]
Absolutely no clue why the Windows build seems to have kicked in with this change @inikem @vlsi and it's finding a lot of flakes - Murphy's all over this one

**Comment by anuraaga** [comment]
Green :)

File: .github/workflows/continuous-build.yml [hunk 1]
```diff
@@ -18,28 +18,32 @@ jobs:
...
+            testAdditionalJavaVersions: true
             coverage: true
-          - os: ubuntu-latest
-            java: 8
     steps:
       - uses: actions/checkout@v2
         with:
           submodules: true
-      - uses: actions/setup-java@v1
+      - id: setup-java-8
```
Comment (iNikem) [inline .github/workflows/continuous-build.yml:29]: If gradle is supposed to downloaded needed toolchain, why do we provide it manually?

Comment (vlsi) [reply .github/workflows/continuous-build.yml:29]: Java is required to launch Gradle itself.

File: examples/gradle/wrapper/gradle-wrapper.properties [hunk 2]
```diff
@@ -1,5 +1,6 @@
 distributionBase=GRADLE_USER_HOME
 distributionPath=wrapper/dists
-distributionUrl=https\://services.gradle.org/distributions/gradle-6.6-all.zip
+distributionUrl=https\://services.gradle.org/distributions/gradle-6.7-all.zip
```
Comment (vlsi) [inline examples/gradle/wrapper/gradle-wrapper.properties:3]: Do you really need `all`?
Can you use `bin` distribution?

File: gradle/wrapper/gradle-wrapper.properties [hunk 3]
```diff
@@ -1,5 +1,6 @@
 distributionBase=GRADLE_USER_HOME
 distributionPath=wrapper/dists
-distributionUrl=https\://services.gradle.org/distributions/gradle-6.6-all.zip
+distributionUrl=https\://services.gradle.org/distributions/gradle-6.7-all.zip
```
Comment (vlsi) [inline gradle/wrapper/gradle-wrapper.properties:3]: Do you really need `all`?
Can you use `bin` distribution?

File: .github/workflows/continuous-build.yml (hunk 1 above)
Comment (iNikem) [reply .github/workflows/continuous-build.yml:29]: For that we have java 11 next to this 8

Comment (anuraaga) [reply .github/workflows/continuous-build.yml:29]: Yeah but setup-java doesn't download Java it just extracts from a local VM disk very quickly. Figure it's worth saving some downloading.

File: examples/gradle/wrapper/gradle-wrapper.properties (hunk 2 above)
Comment (anuraaga) [reply examples/gradle/wrapper/gradle-wrapper.properties:3]: Back in the day -all was required for intellij completion from what I understand. Now I do see gradle compile stubs on first launch with -bin. Does that mean bin is enough for projects, but not plugin authors? If so it mostly makes sense - but I personally found the time it takes for -bin to generate stubs to be way longer (several minutes on a macbook pro) than just downloading -all so not sure it's a savings in practice. What do you think?

Comment (vlsi) [reply examples/gradle/wrapper/gradle-wrapper.properties:3]: Oh, I thought `-all` was generating stubs as well. I did not compare the contents, however, I thought the difference is just source code (*.java files are present in `-all`, and they are not there in `-bin`).

I think `-all` was required earlier for the IDE to navigate to Gradle sources, however, it is not needed now.

Comment (vlsi) [reply examples/gradle/wrapper/gradle-wrapper.properties:3]: https://twitter.com/CedricChampeau/status/1113871652809531392

If you are here, then there's Madrid JUG happening right NOW: https://www.meetup.com/madrid-gug/events/273721823

Modernizing the Groovy build by Cédric Champeau, so you can join and ask `-bin` vs `-all` :)

Comment (anuraaga) [reply examples/gradle/wrapper/gradle-wrapper.properties:3]: I _think_ I noticed slowness with -bin with the generations of stubs that wasn't there with -all but am not sure, I'll need to try a controlled experiments to confirm.

And as much as I [love](https://www.github.com/curioswitch/curiostack/tree/master/tools%2Fgradle-plugins%2Fgradle-curiostack-plugin) Gradle I was supposed to pass out an hour ago and will promptly do so for tomorrow :) Sounds like a fun event for sure though.

**Comment by jkwatson** [comment]
@anuraaga do you think the builds are stable enough to go with this?

**Comment by anuraaga** [comment]
Seems not on Windows yet - let me temporarily remove it from the config since it wasn't activated before anyways for whatever reason and I'll keep on playing with it.

**Comment by anuraaga** [comment]
Switched to -bin

**Review by jkwatson — APPROVED** [review]
Thanks!

File: .circleci/config.yml [hunk 4]
```diff
@@ -28,8 +30,9 @@ jobs:
       # Configure the JVM and Gradle to avoid OOM errors
       _JAVA_OPTIONS: "-Xmx1g"
       GRADLE_OPTS: "-Dorg.gradle.daemon=false -Dorg.gradle.workers.max=2"
+      ORG_GRADLE_PROJECT_testAdditionalJavaVersions: true
     docker:
-      - image: circleci/openjdk:8-jdk
+      - image: circleci/openjdk:11.0.6-jdk-buster
```
Comment (pavolloffay) [inline .circleci/config.yml:35]: Are we sure that this image is being used? The line 25 defines executor type as machine which is VM and I think as it is applied at root all jobs will use it.

Comment (anuraaga) [reply .circleci/config.yml:35]: Right I found out it's not being used. I'm going to be deleting CircleCI build anyways :)

File: build.gradle [hunk 5]
```diff
@@ -169,15 +169,32 @@ configure(opentelemetryProjects) {
         mavenLocal()
     }
 
-    sourceCompatibility = 1.8
-    targetCompatibility = 1.8
```
Comment (tylerbenson) [inline build.gradle:173]: @anuraaga I think by removing this, Intellij doesn't configure the language level correctly.

**Pull request merged**